File: lessons/dsa/complexity-as-a-budget/examples/measure_scans.py
"""Count the steps two duplicate checks spend on the same input.

The point of the lesson is that the step count, not the wall clock, is what
grows with the input. Both functions return the number of comparisons they
performed so the growth can be observed directly.
"""


def steps_for_pairwise_scan(values: list[int]) -> int:
    """Compare every item with every later item. Cost grows with n squared."""
    steps = 0
    for left in range(len(values)):
        for right in range(left + 1, len(values)):
            steps += 1
            if values[left] == values[right]:
                return steps
    return steps


def steps_for_single_scan(values: list[int]) -> int:
    """Read each item once against a set of what was already seen."""
    steps = 0
    seen: set[int] = set()
    for value in values:
        steps += 1
        if value in seen:
            return steps
        seen.add(value)
    return steps

File: lessons/dsa/complexity-as-a-budget/examples/MeasureScans.java
import java.util.HashSet;
import java.util.Set;

/**
 * Count the steps two duplicate checks spend on the same input.
 *
 * <p>The point of the lesson is that the step count, not the wall clock, is
 * what grows with the input, so both methods return the comparisons they made.
 */
public final class MeasureScans {
  private MeasureScans() {}

  /** Compare every item with every later item. Cost grows with n squared. */
  public static int stepsForPairwiseScan(int[] values) {
    int steps = 0;
    for (int left = 0; left < values.length; left++) {
      for (int right = left + 1; right < values.length; right++) {
        steps++;
        if (values[left] == values[right]) {
          return steps;
        }
      }
    }
    return steps;
  }

  /** Read each item once against a set of what was already seen. */
  public static int stepsForSingleScan(int[] values) {
    int steps = 0;
    Set<Integer> seen = new HashSet<>();
    for (int value : values) {
      steps++;
      if (!seen.add(value)) {
        return steps;
      }
    }
    return steps;
  }
}
